### PR TITLE
[docs] Update broadcasting and cuda semantics notes

### PR DESCRIPTION
**Pull request by zou3519**
cc @gchanan

### CODE REVIEW

File: docs/source/notes/broadcasting.rst [hunk 1]
```diff
@@ -19,26 +19,26 @@ Two tensors are "broadcastable" if the following rules hold:
...
 
-    >>> x=torch.FloatTensor(5,7,3)
-    >>> y=torch.FloatTensor(5,7,3)
+    >>> x=torch.empty(5,7,3)
+    >>> y=torch.empty(5,7,3)
     # same shapes are always broadcastable (i.e. the above rules always hold)
 
-    >>> x=torch.FloatTensor()
-    >>> y=torch.FloatTensor(2,2)
+    >>> x=torch.empty(( ,0))
```
Comment (gchanan) [inline docs/source/notes/broadcasting.rst]: this looks wrong.

File: docs/source/notes/broadcasting.rst [hunk 2]
```diff
@@ -19,26 +19,26 @@ Two tensors are "broadcastable" if the following rules hold:
...
+    >>> y=torch.empty((2,2))
```
Comment (gchanan) [inline docs/source/notes/broadcasting.rst]: why is this a tuple when all the other ones aren't?

File: docs/source/notes/cuda.rst [hunk 3]
```diff
@@ -12,35 +12,47 @@ However, once a tensor is allocated, you can do operations on it irrespective
 of the selected device, and the results will be always placed in on the same
 device as the tensor.
 
-Cross-GPU operations are not allowed by default, with the only exception of
-:meth:`~torch.Tensor.copy_`. Unless you enable peer-to-peer memory access, any
-attempts to launch ops on tensors spread across different devices will raise an
-error.
+Cross-GPU operations are not allowed by default, with the exception of
+:meth:`~torch.Tensor.copy_` and other methods like copy like
```
Comment (gchanan) [inline docs/source/notes/cuda.rst]: this sentence looks wrong.  Did you mean other methods with copy like functionality, such as ...?  (don't use like so many times)

File: docs/source/notes/cuda.rst [hunk 4]
```diff
@@ -12,35 +12,47 @@ However, once a tensor is allocated, you can do operations on it irrespective
...
+        b = torch.tensor([1., 2.]).cuda()
+        # a.device.index == b.device.index == 1
+
+        # You can also use ``Tensor.to`` to transfer a tensor:
+        b2 = torch.tensor([1., 2.]).to(device=cuda)
+        b.device.index == b2.device.index == 1
 
         c = a + b
-        # c.get_device() == 1
+        # c.device.index == 1
```
Comment (gchanan) [inline docs/source/notes/cuda.rst]: I'm not really sure that showing the device index is more useful than just showing the device.

File: docs/source/notes/cuda.rst [hunk 5]
```diff
@@ -12,35 +12,47 @@ However, once a tensor is allocated, you can do operations on it irrespective
...
 
-        # even within a context, you can give a GPU id to the .cuda call
-        d = torch.randn(2).cuda(2)
-        # d.get_device() == 2
+        # even within a context, you can specify the device
+        # (or give a GPU index to the .cuda call)
+        d = torch.randn(2, device=cuda2)
+        e = torch.randn(2).to(device=cuda2)
+        f = torch.randn(2).cuda(2)
+        # d.device.index == e.device.index == f.device.index == 2
```
Comment (gchanan) [inline docs/source/notes/cuda.rst]: you can also pass a device to cuda; again, I don't see the advantage of jumping back and forth between index and device.

File: docs/source/notes/broadcasting.rst (hunk 2 above)
Comment (fmassa) [reply docs/source/notes/broadcasting.rst]: I'd maybe encourage writing them all as a tuple, because this is what `numpy.empty` expects?

File: docs/source/notes/cuda.rst [hunk 6]
```diff
@@ -79,8 +91,9 @@ relative order, unless explicit synchronization functions (such as
 :meth:`~torch.cuda.synchronize` or :meth:`~torch.cuda.Stream.wait_stream`) are
 used.  For example, the following code is incorrect::
 
+    cuda = torch.device('cuda')
     s = torch.cuda.stream()  # Create a new stream.
-    A = torch.cuda.FloatTensor(100, 100).normal_(0.0, 1.0)
+    A = torch.empty((100, 100), dtype=cuda).normal_(0.0, 1.0)
```
Comment (fmassa) [inline docs/source/notes/cuda.rst]: I think you meant `device=cuda`?

File: docs/source/notes/cuda.rst [hunk 7]
```diff
@@ -12,35 +12,47 @@ However, once a tensor is allocated, you can do operations on it irrespective
...
-        # z.get_device() == 0
+        # z.device is device(type='cuda', index=0)
 
-        # even within a context, you can give a GPU id to the .cuda call
-        d = torch.randn(2).cuda(2)
-        # d.get_device() == 2
+        # even within a context, you can specify the device
+        # (or give a GPU index to the .cuda call)
+        d = torch.randn(2, device=cuda2)
+        e = torch.randn(2).to(device=cuda2)
```
Comment (gchanan) [inline docs/source/notes/cuda.rst]: nit: these last two cases don't need the kwarg, it's a little verbose.

File: docs/source/notes/cuda.rst [hunk 8]
```diff
@@ -122,8 +135,10 @@ the initial hidden state of a recurrent neural network.
 The first step is to determine whether the GPU should be used or not. A common
 pattern is to use Python's ``argparse`` module to read in user arguments, and
 have a flag that can be used to disable CUDA, in combination with
-:meth:`~torch.cuda.is_available`. In the following, ``args.cuda`` results in a
-flag that can be used to cast tensors and modules to CUDA if desired::
+:meth:`~torch.cuda.is_available`. In the following, ``args.device`` results in a
+:class:`torch.device` object that can be used to cast tensors to CPU or CUDA.
```
Comment (gchanan) [inline docs/source/notes/cuda.rst]: nit: I would say "move" tensors instead of "cast" tensors.

File: docs/source/notes/cuda.rst [hunk 9]
```diff
@@ -132,59 +147,88 @@ flag that can be used to cast tensors and modules to CUDA if desired::
...
+This can be used in a number of cases to produce device agnostic code. Below
+is an example when using a dataloader:
 
-    dtype = torch.cuda.FloatTensor
+::
+
+    cuda0 = torch.device('cuda:0')  # CUDA GPU 0
     for i, x in enumerate(train_loader):
-        x = Variable(x.type(dtype))
+        x = x.to(device=cuda0)
```
Comment (gchanan) [inline docs/source/notes/cuda.rst]: nit: you don't need device kwarg right?

**Review by gchanan — APPROVED** [review]

File: docs/source/notes/cuda.rst [hunk 10]
```diff
@@ -132,59 +147,88 @@ flag that can be used to cast tensors and modules to CUDA if desired::
     parser.add_argument('--disable-cuda', action='store_true',
                         help='Disable CUDA')
     args = parser.parse_args()
-    args.cuda = not args.disable_cuda and torch.cuda.is_available()
+    args.device = None
```
Comment (apaszke) [inline docs/source/notes/cuda.rst]: Why assign `None` to it when you could have just made it `torch.device('cpu')` by default.
